### PR TITLE
Individual document sharing with permissions

### LINKED ISSUE #2359: Individual document sharing
We are exploring enabling individual document sharing within the team, there are a couple of use-cases for this that have come up over time:

- Sharing a draft with a colleague to get feedback or help contributing (cc @matbrgz)
- Sharing an individual document from a private collection for a one-off purpose. For example, imagine there is a "founders" collection that contains sensitive information but you want to share a single document with an EA

The key UX consideration is that this does not become a go-to method of sharing information that should be managed through collection permissions. All too often on other platforms you'll find that you were individually shared a bunch of documents that should have been shared within context, this makes them difficult to find and makes access control auditing difficult.

### Suggested limitations

Documents that have been shared with you…

- will **not** appear in the sidebar. This is key for performance and preventing the above ^.
- will **not** expose their collection or parent document info
- can appear in search results
- can have `read` or `read_write` permissions when shared

### API design

TODO

### UI design

TODO

### PR DESCRIPTION
**Pull request by apoorv-mishra**
Closes #2359 

- [x] `document.add_user`
- [x] `documents.remove_user`
- [x] `documents.shared_with_me`
- [x] Update `documents.search` to also return documents _shared with me_

Fixes:

- [x] `documents.info` returning 403 for the shared doc
- [x] "Deleted Collection" shown in breadcrumbs for individually shared document from a private collection
- [x] _Publish_ and _Duplicate_ actions shown for individually shared doc from a private collection
- [x] Creator unable to publish a draft doc

### CODE REVIEW

**Comment by milotiger** [comment]
We did not have the UI for this PR right? I would love to be an early tester for this feature

**Comment by apoorv-mishra** [comment]
It's a work in progress... 🚧

**Comment by milotiger** [comment]
ok man, really looking forward to this

**Comment by apoorv-mishra** [comment]
- Still, too many issues to fix. Could be classified into three broad categories(not an exhaustive list):
	- Functional
		- At the moment, end of end flow seems to be working fine and there aren't, at least to my knowledge, any *frustrating* issues which would halt the flow. That being said, the functionality is still quite restrictive - only working for documents in a private collection. We also need to be able to facilitate this for drafts in general.
		- Most critical is to watch out for permissions fallout, if any, particularly ones affecting the collection structure as a whole. Example, invited user being able to move or publish the shared doc.
	- UX
		- One is to make it work on smaller devices. At this moment, the list seems to be problematic there.
		- UI fixes likes list hiding inside the share popover(might not be very important rn), limiting height of the list or even the appearance of the invited members list(if too many)
	- Code
		- Cleaning things up! DRYing, example, the `SharedWithMe` component
		  shares a lot with `Starred` component.  
		- `document.*` endpoints vs `userMemberships.*` endpoints - favoring `documents.shared_with_me` or `userMemberships.list`.
		
I believe the functional ones would be most important, followed by UX and Code related ones. Fixed some, but tons remaining. Therefore, **opening up for an initial review from @tommoor  to prioritize fixes.**

File: app/scenes/CollectionPermissions/components/InputMemberPermissionSelect.tsx [hunk 1]
```diff
@@ -13,20 +13,33 @@ export default function InputMemberPermissionSelect(
...
-          label: t("View and edit"),
-          value: CollectionPermission.ReadWrite,
-        },
-        {
-          label: t("Admin"),
-          value: CollectionPermission.Admin,
-        },
-      ]}
+      options={
+        props.isAdminPermissionSupported
```
[the remark this answers is not in the document]
Comment (apoorv-mishra) [reply app/scenes/CollectionPermissions/components/InputMemberPermissionSelect.tsx:17]: Yeah, it was just a _make it work_ , temporary thingy.

File: server/models/helpers/SearchHelper.ts [hunk 2]
```diff
@@ -342,20 +344,66 @@ export default class SearchHelper {
         : '"publishedAt" IS NOT NULL'
     }
   `;
-    const selectSql = `
+    const selectSql =
```
[the remark this answers is not in the document]
Comment (apoorv-mishra) [reply server/models/helpers/SearchHelper.ts]: No it was a `UNION`. Anyways, modified the where clause altogether now.

**Comment by tommoor** [comment]
Okay, pretty sure this is good to go – there are a number of post-merge follow up tasks but none are blocking an initial rollout. 

Just want to do one last pass through the test suite and make sure all the main negative cases are covered.

**Review by apoorv-mishra — COMMENTED** [review]
image

Both _back_ and _Done_ buttons achieve same result?

**Review by apoorv-mishra — COMMENTED** [review]
Screenshot 2024-01-30 at 19 20 06

A tooltip telling that this copies the doc link would help?

**Comment by apoorv-mishra** [comment]
Screenshot 2024-01-30 at 19 24 33

On smaller device, the input box shifts towards right when focused, to give space to the back button. The back button could be moved inside the input boundary, similar to copy link button?

**Comment by apoorv-mishra** [comment]
https://github.com/outline/outline/assets/15694566/f1ace0d0-04b2-48ba-b5f9-9ecbcdc5a604

`documents.add_user` getting fired in the other user's session.

**Comment by apoorv-mishra** [comment]
https://github.com/outline/outline/assets/15694566/1f132f31-3e6f-4c02-af39-7c240e79783a

Can't really see what's being typed in the custom link box, just the cursor blinking.

**Comment by apoorv-mishra** [comment]
https://github.com/outline/outline/assets/15694566/fd2e0461-b7cc-4199-bb36-1f7c78e05447

Can hide the workspace members who already have access. Might be tricky since we're currently relying on `UserMembership` for access here while this is because of the parent collection's permission.

**Pull request merged**